tensorstore/kvstore/s3/s3_request_builder.cc

```diff
@@ -0,0 +1,253 @@
+  );
+}
+
+static constexpr char kHostHeader[] = "host: ";
+static constexpr char kAmzContentSha256Header[] = "x-amz-content-sha256: ";
+static constexpr char kAmzSecurityTokenHeader[] = "x-amz-security-token: ";
+
+} // namespace
+
+
+HttpRequest S3RequestBuilder::BuildRequest(
+  std::string_view host,
+  const S3Credentials & credentials,
+  std::string_view aws_region,
+  std::string_view payload_sha256_hash,
+  const absl::Time & time) {
+
+    builder_.AddHeader(absl::StrCat(kHostHeader, host));
+    builder_.AddHeader(absl::StrCat(kAmzContentSha256Header, payload_sha256_hash));
+    builder_.AddHeader(absl::FormatTime("x-amz-date: %Y%m%dT%H%M%SZ", time, absl::UTCTimeZone()));
+
+    // Add deferred query parameters in sorted order for AWS4 signature requirements
+    std::stable_sort(std::begin(query_params_), std::end(query_params_));
+    for (const auto& [k, v] : query_params_) {
+      builder_.AddQueryParameter(k, v);
+    }
+
+    // If anonymous, it's unnecessary to construct the Authorization header
+    if (credentials.IsAnonymous()) {
+      return builder_.BuildRequest();
+    }
+
+    // Add AWS Session Token, if available
+    // https://docs.aws.amazon.com/AmazonS3/latest/userguide/RESTAuthentication.html#UsingTemporarySecurityCredentials
+    if (!credentials.session_token.empty()) {
+      builder_.AddHeader(absl::StrCat(kAmzSecurityTokenHeader, credentials.session_token));
+    }
+
+    auto request = builder_.BuildRequest();
+
+    // Create sorteded AWS4 signing headers
+    std::vector<std::pair<std::string, std::string>> signed_headers;
+
+    for(const auto & header_str: request.headers) {
+      auto header = std::string_view(header_str);
+      auto pos = header.find(':');
+      assert(pos != std::string::npos);
+      auto key = absl::AsciiStrToLower(absl::StripAsciiWhitespace(header.substr(0, pos)));
+      auto value = absl::StripAsciiWhitespace(header.substr(pos + 1));
+      signed_headers.push_back({std::string(key), std::string(value)});
+    }
+
+    std::stable_sort(std::begin(signed_headers), std::end(signed_headers));
+
+    canonical_request_ = CanonicalRequest(request.url, request.method,
+                                              payload_sha256_hash, signed_headers);
+    signing_string_ = SigningString(canonical_request_, aws_region, time);
+    signature_ = Signature(credentials.secret_key, aws_region, signing_string_, time);
+    auto auth_header = AuthorizationHeader(credentials.access_key, aws_region, signature_,
+                                          signed_headers, time);
+
+    ABSL_LOG_IF(INFO, TENSORSTORE_INTERNAL_S3_LOG_AWS4)
+        << "Canonical Request\n" << canonical_request_;
+    ABSL_LOG_IF(INFO, TENSORSTORE_INTERNAL_S3_LOG_AWS4)
+        << "Signing String\n" << signing_string_;
+    ABSL_LOG_IF(INFO, TENSORSTORE_INTERNAL_S3_LOG_AWS4)
+        << "Authorization Header\n" << auth_header;
+
+    request.headers.emplace_back(std::move(auth_header));
+    return request;
+  }
```

Comment:
indent.  we should probably have a clang-format somewhere to use.

Reply:
I've added a reminder to lint with clang-format --style=google in the TODO.